language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs

```diff
@@ -162,17 +169,147 @@ getBuiltinFoldList = do
+--
+-- @list (integer s)@ appears several times in types,
+-- so this is not really a definition with unique names.
+getBuiltinEnumFromTo :: Quote (Term TyName Name ())
+getBuiltinEnumFromTo = do
+    fix         <- getBuiltinFix
+    succInteger <- getBuiltinSuccInteger
+    unit        <- getBuiltinUnit
+    ifThenElse  <- getBuiltinIf
+    list        <- getBuiltinList
+    nil         <- getBuiltinNil
+    cons        <- getBuiltinCons
+    s <- freshTyName () "s"
+    ss  <- freshName () "ss"
+    n   <- freshName () "n"
+    m   <- freshName () "m"
+    rec <- freshName () "rec"
+    n'  <- freshName () "n'"
+    u   <- freshName () "u"
+    let eqInteger  = Constant () $ BuiltinName () EqInteger
+        int = TyApp () (TyBuiltin () TyInteger) $ TyVar () s
+        RecursiveType _ listInt =
+            holedToRecursive $ holedTyApp list int
+    return
+        . TyAbs () s (Size ())
+        . LamAbs () ss (TyApp () (TyBuiltin () TySize) $ TyVar () s)
+        . LamAbs () n int
+        . LamAbs () m int
+        . mkIterApp (mkIterInst fix [int, listInt])
+        $ [   LamAbs () rec (TyFun () int listInt)
+            . LamAbs () n' int
+            . mkIterApp (TyInst () ifThenElse listInt)
+            $ [ mkIterApp (TyInst () eqInteger $ TyVar () s)
+                    [ Var () n'
+                    , Var () m
+                    ]
+              , LamAbs () u unit $ TyInst () nil int
+              , LamAbs () u unit $ mkIterApp (TyInst () cons int)
+                    [ Var () n'
+                    ,   Apply () (Var () rec)
+                      . mkIterApp (TyInst () succInteger (TyVar () s))
+                      $ [ Var () ss
+                        , Var () n'
+                        ]
+                    ]
+              ]
+          , Var () n
+          ]
+
 -- |  'sum' as a PLC term.
 --
--- > /\(s :: *) -> foldLst {integer s} {integer s} (addInteger {s}) s!0
+-- > /\(s :: *) -> \(ss : size s) ->
+-- >     foldList {integer s} {integer s} (addInteger {s}) (resizeInteger {1} {s} ss 1!0)
+getBuiltinSum :: Quote (Term TyName Name ())
+getBuiltinSum = do
+    foldList <- getBuiltinFoldList
+    s  <- freshTyName () "s"
+    ss <- freshName () "ss"
+    let sv  = TyVar () s
+        int = TyApp () (TyBuiltin () TyInteger) sv
+        add = TyInst () (Constant () (BuiltinName () AddInteger)) sv
+    return
+        . TyAbs () s (Size ())
+        . LamAbs () ss (TyApp () (TyBuiltin () TySize) sv)
+        . mkIterApp (mkIterInst foldList [int, int])
+        $ [ add
+          , makeDynamicBuiltinInt sv (Var () ss) 0
+          ]
+
+-- |  'product' as a PLC term.
 --
--- TODO: once sizes are added, make the implementation match the comment (which is wrong).
-getBuiltinSum :: Natural -> Quote (Term TyName Name ())
-getBuiltinSum s = do
+-- > /\(s :: *) -> \(ss : size s) ->
+-- >     foldList {integer s} {integer s} (addInteger {s}) (resizeInteger {1} {s} ss 1!0)
+getBuiltinProduct :: Quote (Term TyName Name ())
```

Comment:
Oops, uses `addInteger` rather than `multiplyInteger`. I'll fix that in the subsequent PR.